library/core/src/iter/adapters/flatten.rs

```diff
@@ -348,6 +348,40 @@ where
 
         init
     }
+
+    #[inline]
+    #[rustc_inherit_overflow_checks]
+    fn advance_by(&mut self, n: usize) -> Result<(), usize> {
+        let mut rem = n;
+        loop {
+            if let Some(ref mut front) = self.frontiter {
+                match front.advance_by(rem) {
+                    ret @ Ok(_) => return ret,
+                    Err(advanced) => rem -= advanced,
+                }
+            }
+            self.frontiter = match self.iter.next() {
+                Some(iterable) => Some(iterable.into_iter()),
+                _ => break,
+            }
+        }
+
+        self.frontiter = None;
+
+        if let Some(ref mut back) = self.backiter {
+            if let Err(advanced) = back.advance_by(rem) {
+                rem -= advanced
+            }
+        }
+
+        if rem > 0 {
+            return Err(n - rem);
+        }
```

Comment:
What if `back.advance_by(rem)` returns `Ok(())`? You'll either end up returning an `Err` or you'll wrongly set `self.backiter` to `None`. Note that the implementation of `advance_back_by` correctly handles this by matching on its return value.